src/main/java/org/elasticsearch/index/translog/fs/FsTranslogFile.java

```diff
@@ -59,27 +74,77 @@ public static Type fromString(String type) throws ElasticsearchIllegalArgumentEx
         }
     }
 
-    long id();
 
-    int estimatedNumberOfOperations();
+    public abstract Translog.Location add(BytesReference data) throws IOException;
+
+    public abstract void reuse(FsTranslogFile other) throws TranslogException;
+
+    public abstract void updateBufferSize(int bufferSize) throws TranslogException;
+
+    public abstract void sync() throws IOException;
+
+    public abstract boolean syncNeeded();
+
+    @Override
+    public FsChannelSnapshot snapshot() {
+        return new FsChannelSnapshot(immutableReader());
+    }
+
+    /**
+     * returns a new reader that follows the current writes (most importantly allows making
+     * repeated snapshots that includes new content)
+     */
+    public FsChannelReader reader() {
+        channelReference.incRef();
+        boolean success = false;
+        try {
+            FsChannelReader reader = new InnerReader(this.id, channelReference);
+            success = true;
+            return reader;
+        } finally {
+            if (!success) {
+                channelReference.decRef();
+            }
+        }
+    }
+
+
+    /** returns a new immutable reader which only exposes the current written operation * */
+    abstract public FsChannelImmutableReader immutableReader();
 
-    long translogSizeInBytes();
+    boolean assertBytesAtLocation(Translog.Location location, BytesReference expectedBytes) throws IOException {
+        ByteBuffer buffer = ByteBuffer.allocate(location.size);
+        readBytes(buffer, location.translogLocation);
+        return new BytesArray(buffer.array()).equals(expectedBytes);
+    }
 
-    Translog.Location add(BytesReference data) throws IOException;
 
-    byte[] read(Translog.Location location) throws IOException;
+    class InnerReader extends FsChannelReader {
```

Comment:
this needs access to the inner workings of the parent class. I explained in the new java docs :)